### PR TITLE
Fix infinite recursion in API BC test

### LINKED ISSUE #104012: DISABLED test_public_api_surface (__main__.TestFXAPIBackwardCompatibility)
Platforms: asan, linux, mac, macos, rocm, win, windows

This test was disabled because it is failing in CI. See [recent examples](https://hud.pytorch.org/flakytest?name=test_public_api_surface&suite=TestFXAPIBackwardCompatibility) and the most recent trunk [workflow logs](https://github.com/pytorch/pytorch/runs/14450059788).

Over the past 3 hours, it has been determined flaky in 117 workflow(s) with 351 failures and 117 successes.

**Debugging instructions (after clicking on the recent samples link):**
DO NOT ASSUME THINGS ARE OKAY IF THE CI IS GREEN. We now shield flaky tests from developers so CI will thus be green but it will be harder to parse the logs.
To find relevant log snippets:
1. Click on the workflow logs linked above
2. Click on the Test step of the job so that it is expanded. Otherwise, the grepping will not work.
3. Grep for `test_public_api_surface`
4. There should be several instances run (as flaky tests are rerun in CI) from which you can study the logs.

Test file path: `test_fx.py`

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

### PR DESCRIPTION
**Pull request by aorenste**
```
python test/test_fx.py -k test_public_api_surface
```
was failing with a complaint about infinite recursion. Fixed that and then marked the two API changes from #123681 as private (for `get_example_value`) and backward compatible (for `insert_deferred_runtime_asserts`).

Fixes #104012 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125706

### CODE REVIEW

**Comment by aorenste** [comment]
@BoyuanFeng Can you verify that these two entrypoints should be considered backward compatible?

File: torch/fx/passes/runtime_assert.py [hunk 1]
```diff
@@ -14,11 +14,13 @@
...
 from torch.fx.experimental.sym_node import SymNode
 from torch.fx.graph_module import GraphModule
+from torch.fx._compatibility import compatibility
 
 log = logging.getLogger(__name__)
 graph_code_log = torch._logging.getArtifactLogger(__name__, "graph_code")
 
 
+@compatibility(is_backward_compatible=True)
 def get_example_value(node: fx.Node) -> Optional[str]:
```
Comment (ezyang) [inline torch/fx/passes/runtime_assert.py:24]: We can land this as is to unblock but this function should be private

Comment (aorenste) [reply torch/fx/passes/runtime_assert.py:24]: I need to update anyway since, as usual, I forgot to run lintrunner before submitting...

**Review by BoyuanFeng — APPROVED** [review]

**Pull request closed without merging**